azavea/pfb-network-connectivity
# Send status updates from analysis jobs

Makes analysis jobs update their status in the management app.
Pieces::
- Adds an `AnalysisJobStatusUpdate` model to track updates for a given job. It's foreign keyed to the AnalysisJob and timestamped.
- Adds a management command, `update_status` that takes `job_id`, `status`, `step`, and `message` arguments and creates an `AnalysisJobStatusUpdate` instance.
- Pulls the whole Django app into the analysis container, so that the analysis can run the management command to update its status (rather than having to hit an endpoint). See below for some notes on this.
   - Passes all the environment variables that the Django app needs to run as container environment overrides when a job is made, since the job container now needs to be able to run the django app.
- Adds a `utils.sh` script to the analysis scripts directory with an `update_status` command that wraps the management command, so other scripts can source the file and use it.
- Adds `update_status` calls at what seemed like useful spots throughout the analysis.
- Moves the analysis container into docker compose.
- Moves the analysis from `./pfb-analysis` to `./src/analysis`. I think this makes sense in general, but the immediate issue it solves is that with the project root directory as the nearest shared ancestor of the analysis and the django app, the Docker build context had to be `./`, which meant it pulled in everything, including `./data`.

Incidental changes:
- Added `end=''` to the `print` statement in `update_job_definition.py`. This might be a platform-dependent thing, but it was adding a `\r` that ended up in the `PFB_AWS_BATCH_JOB_DEFINITION_NAME_REVISION` value in `staging-pfb-config-us-east-1.tfplan`, which made terraform mad.
- Upgraded terraform in `group_vars/all.example` to 0.8.8.  Jenkins evidently got updated, so my 0.8.7 terraform refused to mess with a configuration created by a newer version.

Notes:
- There are pros and cons to the "include Django" strategy rather than the endpoint strategy
   - Pros:
      - Slightly fewer moving pieces in the code path
      - If we end up needing to dump large quantities of data from the analysis, we'll be set up to save it right into the app's database.
   - Cons:
      - More moving pieces in the deployment/configuration setup
      - Passing the environment through is a little weird and could be a source of bugs during changes or refactors.
- There's a `message` property on `AnalysisJobStatusUpdate` that I didn't end up using, but I didn't delete it because I could imagine it being useful for error statuses.  Easy to remove, though, if we don't want it.
- The "install dependencies" step of the analysis container build already took a while, and now it takes slightly longer. It also now needs the Django app's `requirements.txt`, but to avoid re-building the dependencies layer every time the Django code changes, I set it up to copy only the `requirements.txt` first and then copy the rest of the app in a later step.

**To test**
Unfortunately passing the environment to the job on creation means it won't work to create jobs in your local instance to run on staging.  Perhaps we could set up something easier to use, but what I did to test this was:
- Deploy to staging (the usual cibuild/cipublish/infra plan/infra apply)
- Log into the staging task (`ssh -A` to bastion, ssh to instance, `docker exec -ti <container id> bash`)
- Run migrations (`./manage.py migrate`)
- In `shell_plus`, do
```
user = PFBUser.objects.first()
gtown = Neighborhood.objects.get(name='germantown-philadelphia-pa')
job = AnalysisJob.objects.create(neighborhood=gtown, created_by=user, modified_by=user)
job.run()
```
Then you can see the updates with `print '\n'.join([' '.join((str(u.timestamp), str(u.status), str(u.step))) for u in job.status_updates.all()])`
![image](https://cloud.githubusercontent.com/assets/6598836/24110254/5c742dca-0d69-11e7-9501-66e1d91af85c.png)


Connects #145.  Also connects #126.
